serialize policy repository update events

ianvernon opened the pull request:
This originally was part of 
https://github.com/cilium/cilium/pull/6943 .

This PR serves to serialize the various types of events which change the state of the policy repository to avoid race-y behavior from occurring (from FQDN policy updates, for instance). It does this via using a new structure, the EventQueue.


The EventQueue package implements a queue-based system for event processing in a generic fashion in a first-in, first-out manner. An EventQueue may be closed, in which case all events which are queued up, but have not been processed, will be cancelled (i.e., not ran). It is guaranteed that no events will be scheduled onto an EventQueue after it has been closed; if any event is attempted to be scheduled onto an EventQueue after it has been closed, it will be cancelled immediately. For any event to be processed by the EventQueue, it must implement the EventHandler interface. This allows for different types of events to be processed by anything which chooses to utilize an EventQueue.

To ensure correctness in ordering these events, and for future work, two EventQueues are added to the Repository type:

* RepositoryChangeQueue: this serializes policy repository changes (update, delete, insert). This ensures that the operations associated with updating the policy repository are serialized, without the entirety of the operations holding the repository lock.
* RuleReactionQueue : this serializes the 'reacting' to events based on policy changes without blocking the policy API. This is not used currently, and will be utilized in future work (#6943).

Both of these queues are buffered up to 100 events each. After this point, the API will block. Their length is configurable via the configuration option `policy-queue-size`. 

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7445)
<!-- Reviewable:end -->

Commit:
add eventqueue package

This package implements a queue-based system for event processing in a generic
fashion in a first-in, first-out manner. Only one event is able to be consumed
at a time per event queue. This is useful in the case for ensuring that certain
types of events do not run concurrently. An EventQueue may be closed, in which
case all events which are queued up, but have not been processed, will be
cancelled (i.e., not ran). It is guaranteed that no events will be scheduled
onto an EventQueue after it has been closed; if any event is attempted to be
scheduled onto an EventQueue after it has been closed, it will be cancelled
immediately. For any event to be processed by the EventQueue, it must implement
the `EventHandler` interface. This allows for different types of events to be
processed by anything which chooses to utilize an `EventQueue`.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
daemon: update documentation for PolicyAdd

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
serialize policy repository update eventss

The ordering of when all work associated with a given policy update
occurs is important, as the policy repository is modified by the DNS proxy.

Two EventQueues are added: one, which serializes the processing of policy
repository changes (add, delete, update), and one which serializes the
'reacting' to rule updates (will be used in the future).

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
daemon: fix race in unit tests

Now that `PolicyAdd` signals that it is done before updating of prefix
lengths is finished, update the unit test accordingly to use
`testutils.WaitUntil` to wait for prefix lengths to be correct instead of
performing the assertion once and bailing out before the operation may have
completed.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: make queue sizes in policy repository configurable

The size of the queues for policy-related events, and reaction events can now be
customized by end-users, with the "policy-queue-size" option. The default value
is 100.

Signed-off by: Ian Vernon <ian@cilium.io>

joestringer commented on pkg/eventqueue/eventqueue.go:
```diff
@@ -0,0 +1,263 @@
+
+// Run consumes events that have been queued for this EventQueue. It
+// is presumed that the eventQueue is a buffered channel with a length of one
+// (i.e., only one event can be processed at a time). All business logic for
```
nit: I don't think this assumption still holds?

ianvernon replied:
I thought I had caught this one, sorry for the noise. Will update.

joestringer commented on pkg/policy/repository.go:
```diff
@@ -38,12 +39,28 @@ type Repository struct {
+	// need to occur after updating the state of the policy repository. This
+	// can include queueing endpoint regenerations, policy revision increments
+	// for endpoints, etc.
+	RuleReactionQueue *eventqueue.EventQueue
```
nit: Strictly speaking this part is no-op right now so should go into the other PR, but I expect at this point that both PRs should go in anyway (modulo review) so it's not that big of a deal to have it here.

joestringer reviewed (approved):
Very minor nits, LGTM overall. 

Still looking for some documentation around the magic numbers for the queues here about some of the considerations for how to decide on a good number for it.

nebril commented on daemon/policy.go:
```diff
@@ -178,14 +179,58 @@ type AddOptions struct {
+	polAddEvent := eventqueue.NewEvent(p)
+	resChan := d.policy.RepositoryChangeQueue.Enqueue(polAddEvent)
+
+	select {
```
Select is superfluous here, there is only one case.

nebril commented on daemon/policy.go:
```diff
@@ -225,21 +278,29 @@ func (d *Daemon) PolicyAdd(rules policyAPI.Rules, opts *AddOptions) (newRev uint
+		return
 	}
 
+	// No errors past this point!
```
Can you elaborate on this comment? Do you not expect any errors past this point, or any error past this point will crash the daemon?

ianvernon replied:
Went ahead and removed it since it doesn't add value. 

Basically, there should be no more errors related to policy being added past this point because we've done all operations that could fail.

nebril commented on pkg/eventqueue/eventqueue.go:
```diff
@@ -0,0 +1,263 @@
+	// order for the Event to be successfully processed by the EventQueue.
+	Metadata EventHandler
+
+	// EventResults is a channel on which the results of the event are sent.
```
s/EventResults/eventResults

nebril reviewed (commented):
Looks great overall, small nitpicks left inline.

Commit:
daemon: remove superfluous `select` statements

No functional change is intended here.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
daemon: remove useless comment

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
eventqueue: correct case in documentation

Signed-off by: Ian Vernon <ian@cilium.io>

nebril reviewed (approved):
LGTM overall. I thought if it would be possible to not use `interface{}` as channel type for `eventResults`, but couldn't come up with any reasonable abstraction that would be broad enough to accommodate everyone, but not be `interface{}`.
